lib/services/broker/wsBrokerServer.js

```diff
@@ -0,0 +1,275 @@
+    deferred.reject(new InternalError('Websocket server already started'));
+    return deferred.promise;
+  }
+
+  this.server = this.ws(this.options.server, () => {
+    deferred.resolve();
+  });
+
+  this.server.on('connection', clientSocket => {
+    clientSocket.on('message', msg => {
+      msg = JSON.parse(msg);
+
+      switch (msg.action) {
+        // the listen action is called by the client to inform the server it
+        // wants to be notified on the room activity
+        case 'listen':
+          if (this.rooms[msg.room] === undefined) {
+            this.rooms[msg.room] = new CircularList([]);
+          }
+          if (this.rooms[msg.room].getArray().indexOf(clientSocket) === -1) {
+            this.rooms[msg.room].add(clientSocket);
+          }
+          break;
+        case 'unsubscribe':
+          if (this.rooms[msg.room]) {
+            this.rooms[msg.room].remove(clientSocket);
+            if (this.rooms[msg.room].getSize() === 0) {
+              delete this.rooms[msg.room];
+            }
+          }
+          break;
+
+        // broadcast & send can be used in 2 ways:
+        //   1 - (most common): pass a message to the server to a room on which
+        //                      a callback was set to trigger a remote action
+        //   2 - If some other clients have subscribed to the room, they will
+        //       be also notified
+        case 'send':
+          this.dispatch(msg.room, msg.data);
+          this.send(msg.room, msg.data, clientSocket);
+          break;
+        case 'broadcast':
+          this.dispatch(msg.room, msg.data);
+          this.broadcast(msg.room, msg.data, clientSocket);
+          break;
+      }
+    });
+
+    clientSocket.on('close', (code, message) => {
+      _kuzzle.pluginsManager.trigger('log:info', `client disconnected [${code}] ${message}`);
+      removeClient.call(this, clientSocket);
+    });
+
+    clientSocket.on('error', err => {
+      _kuzzle.pluginsManager.trigger('log:error', err);
+    });
+  });
+
+  return deferred.promise;
+};
+
+/**
+ * Broadcasts `data` to *all* clients that have subscribed to the `room` but
+ * the emitter.
+ *
+ * @param {string} room
+ * @param {object} data
+ * @returns {int} Number of clients the message was sent to ; -1 if the room does not exist.
+ */
+WSBrokerServer.prototype.broadcast = function (room, data, emitterSocket) {
+  var clients = 0;
+
+  if (!this.rooms[room]) {
+    return -1;
+  }
+
+  async.each(
+    this.rooms[room].getArray(),
+    (clientSocket, cb) => {
+      try {
+        if (clientSocket === emitterSocket) {
+          return cb();
+        }
+
+        clients++;
+        clientSocket.send(JSON.stringify({
+          room: room,
+          data: data
+        }));
+
+        cb();
+      }
+      catch (err) {
+        cb(err);
+      }
+    },
+    err => {
+      if (err) {
+        clients = -1;
+        _kuzzle.pluginsManager.trigger('log:error', err);
+      }
+    });
+
+  return clients;
```

Comment:
The `clients` variable may not be filled with the real number of clients due to `async.each` usage.

I think we must use the third argument of `async.each(coll, iteratee, [callback])` to resolve the `clients` in a promise 
Or use `Array.forEach` if we have to keep the `broadcast` in a synchrone way.